contrib/submit-protocol/src/App/ProtocolForm.tsx

```diff
@@ -0,0 +1,227 @@
+        return res.json();
+      }).then((body) => {
+        const protocol = jsyaml.safeLoad(body);
+        this.setState({ protocol });
+        this.setJobName(`${source.jobName}_clone_${crypto.randomBytes(4).toString("hex")}`);
+      }).catch((err) => {
+        window.alert(err.message);
+      });
+    }
+  }
+
+  public render() {
+    return (
+      <>
+        <Fabric>
+          <Panel
+            isOpen={this.state.showEditor}
+            isLightDismiss={true}
+            onDismiss={this.closeEditor}
+            type={PanelType.largeFixed}
+            headerText="Protocol YAML Editor"
+          >
+            <MonacoEditor
+              width={800}
+              height={800}
+              value={this.state.protocolYAML}
+              onChange={this.editProtocol}
+              language="yaml"
+              theme="vs-dark"
+              options={{ wordWrap: "on", readOnly: false }}
+            />
+            <div style={{ marginTop: "15px" }}>
+              <PrimaryButton text="Save" onClick={this.saveEditor} style={{ marginRight: "10px" }}/>
+              <DefaultButton text="Discard" onClick={this.discardEditor} />
+            </div>
+          </Panel>
+          <div className="content">
+            <form className="modal-dialog">
+              <div className="modal-content">
+                <div className="modal-header">
+                  <h3 className="Header">
+                    Submit Protocol <small style={{ margin: "10px" }}>PREVIEW</small>
+                  </h3>
+                </div>
+                <div className="modal-body row">
+                  <div className="form-group col-md-8">
+                    <TextField
+                      label="Job Name "
+                      value={this.state.jobName}
+                      onChanged={this.setJobName}
+                      required={true}
+                    />
+                  </div>
+                  <div className="form-group col-md-8">
+                    <Label>Protocol YAML Operation</Label>
+                      <label className="col-md-3" style={{padding: 0}}>
+                        <a type="button" className="btn btn-default">Import</a>
+                        <input
+                          type="file"
+                          className="sr-only"
+                          accept=".yml,.yaml"
+                          onChange={this.importFile}
+                        />
+                      </label>
+                      <DefaultButton text="View/Edit" onClick={this.openEditor} />
+                  </div>
+                </div>
+                <div className="modal-footer" style={{ marginTop: "256px" }}>
+                  <PrimaryButton text="Submit Job" onClick={this.submitProtocol} />
+                </div>
+              </div>
+            </form>
+          </div>
+        </Fabric>
+      </>
+    );
+  }
+
+  private setJobName = (jobName: string) => {
+    const protocol = this.state.protocol;
```

Comment:
React state is immutable. Please use `lodash.cloneDeep` or something else to get a clone.
https://medium.freecodecamp.org/handling-state-in-react-four-immutable-approaches-to-consider-d1f5c00249d5

Reply:
Updated, use [react `update`](https://reactjs.org/docs/update.html).